### ISSUE TITLE
DIRAC bulk jobs submission (parametric jobs)

### ISSUE THREAD

**Issue by fstagni** [ganga-devs/ganga#1350]
This issue is about using the parametric jobs submission of DIRAC (i.e. "bulk submission") for jobs submission speedup. Maybe there's another issue/task somewhere but I can't find it. If there's one, I am sorry for the duplication.

Since DIRAC v6r20p12, I believe there's in place everything that's needed for ganga to submit parametric jobs to DIRAC, which would greatly improve the submission speed of jobs. Some documentation is here: https://dirac.readthedocs.io/en/latest/UserGuide/Tutorials/JobManagementAdvanced/index.html#parametric-jobs

How to invoke this in a "hello world" example:
```
from DIRAC.Interfaces.API.Job import Job
from DIRAC.Interfaces.API.Dirac import Dirac
# or extensions, e.g. from LHCbDIRAC.Interfaces.API.LHCbJob import LHCbJob for LHCb

J = Job()
J.setCPUTime(17800)
J.setInputSandbox('exe-script.py')
J.setParameterSequence("args", ['one', 'two', 'three'])
J.setParameterSequence("iargs", [1, 2, 3])
J.setExecutable("exe-script.py", arguments=": testing %(args)s %(iargs)s", logFile='helloWorld_%n.log')
Dirac().submit(J)
```

Now, there's a limit to the maximum amount of parametric jobs that can be submitted, and this limit HAVE to be discovered before attempting submitting a parametric job, and if maybe split in multiple chunks:
```
from DIRAC.WorkloadManagementSystem.Client.JobManagerClient import JobManagerClient
JobManagerClient().getMaxParametricJobs()
```

**Comment by mesmith75** [comment]
Great! As far as I am aware this has been in place from the ganga side for a long time. I'll have a little test today to make sure it still works as intended.

**Comment by fstagni** [comment]
Well the check on th `maxParametricJobs` is new, so you'd only need to add that.

**Comment by egede** [comment]
@fstagni So the reply from `JobManagerClient().getMaxParametricJobs()` tells how many individual jobs can go into a given parametric job? Is this number something that is changing dynamically? Or can we get this number the first time a Ganga session talks to Dirac and then just cache it for the rest of its sesion (that might last many days in principle).
`

**Comment by fstagni** [comment]
This number MAY change at any time BUT won't change often. To be sure I would get it every time you submit a parametric job, and it's not really a performance issue.

**Comment by mesmith75** [comment]
@fstagni How does this work with inputdata?

**Comment by fstagni** [comment]
Still with `setParameterSequence`:

```
J.setParameterSequence('InputData', lfns, addToWorkflow=True)
```
and for the input sandbox:
```
J.setParameterSequence('InputSandbox', files, addToWorkflow=True)
```

**Comment by mesmith75** [comment]
So ganga still has to split the jobs or is that taken care of by Dirac? Is there a parameter to set for the no. of files per job?

**Comment by egede** [comment]
@fstagni I do not understand syntax for inoutdata and the inputsandbox. So for the `files` above, is that just one long list that DIARC will split, or is it a list of lists? For `files` in the inputfiles, there is no way for Dirac to know if I want a job for each element in the list, or if every job should have every file. If it is a list of lists, then that can be specified, but is that what is expected?

**Comment by mesmith75** [comment]
@egede As far as I can understand it is a list of list - in which case ganga still has to do the splitting of the dataset. The splitting is the bit that takes the time at the moment so the gain will not be that significant.

**Comment by fstagni** [comment]
Sorry for the late reply. As of today splitting of the InputData and of the InputSandbox is left to the user (Ganga). It can/should be in the form of a list of lists, and its `len` has to be the same `len` for each parameter, so it should not exceed `maxParametricJobs`.

I added some info in https://github.com/DIRACGrid/DIRAC/pull/3847

**Comment by egede** [comment]
@fstagni Thanks. @mesmith75 will implement this. However, the main time submission time at the moment is actually taken by the splitting, so the gain in speed will be small for the standard DaVinci type jobs. For jobs without input data (such as toy MC studies) there will be a larger gain.

**Comment by mesmith75** [comment]
@fstagni I am unable to get the `InputSandbox` to work with a list of lists. My job looks like:
```python
j = LHCbJob()

# default commands added by ganga
j.setName('minJob_test__{Ganga_GaudiExec_(0.0)}')
j.setParameterSequence("exepar", ['jobScript/masmith_GaudiExec_Job_0.0_26002155-98cf-4e6e-b1a6-d12182407250_script.py', 'jobScript/masmith_GaudiExec_Job_0.1_26002155-98cf-4e6e-b1a6-d12182407250_script.py'])
j.setParameterSequence("logpar", ['Ganga_GaudiExec.log', 'Ganga_GaudiExec.log'])
j.setExecutable("%(exepar)s", logFile="%(logpar)s")
j.setParameterSequence("InputSandbox", [['/afs/cern.ch/user/m/masmith/2910_LFNs.txt', '/afs/cern.ch/user/m/masmith/2912_LFNs.txt'], ['/afs/cern.ch/user/m/masmith/2931_LFNs.txt', '/afs/cern.ch/user/m/masmith/2935_LFNs.txt']], addToWorkflow=True)
j.setOutputSandbox(['summary.xml', '*.root', '__postprocesslocations__'])
j.setParameterSequence("InputData", [['/lhcb/MC/2012/ALLSTREAMS.DST/00042092/0000/00042092_00000057_2.AllStreams.dst'], ['/lhcb/MC/2012/ALLSTREAMS.DST/00042092/0000/00042092_00000091_2.AllStreams.dst']], addToWorkflow=True)
```
but I get an error about the sandbox:
```python
Job submission failure Var InputSandbox : value is opened with " but is not closed
{'CallStack': ['  File "/opt/dirac/pro/Linux_x86_64_glibc-2.12/lib/python2.7/threading.py", line 774, in __bootstrap\n    self.__bootstrap_inner()\n', '  File "/opt/dirac/pro/Linux_x86_64_glibc-2.12/lib/python2.7/threading.py", line 801, in __bootstrap_inner\n    self.run()\n', '  File "/opt/dirac/pro/DIRAC/Core/Utilities/ThreadPool.py", line 111, in run\n    oJob.process()\n', '  File "/opt/dirac/pro/DIRAC/Core/Utilities/ThreadPool.py", line 161, in process\n    self.__jobResult = self.__jobFunction( *self.__jobArgs, **self.__jobKwArgs )\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/private/Service.py", line 352, in _processInThread\n    result = self._processProposal( trid, proposalTuple, handlerObj )\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/private/Service.py", line 512, in _processProposal\n    result = self._executeAction( trid, proposalTuple, handlerObj )\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/private/Service.py", line 532, in _executeAction\n    return handlerObj._rh_executeAction( proposalTuple )\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/RequestHandler.py", line 123, in _rh_executeAction\n    retVal = self.__doRPC(actionTuple[1])\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/RequestHandler.py", line 240, in __doRPC\n    return self.__RPCCallFunction(method, args)\n', '  File "/opt/dirac/pro/DIRAC/Core/DISET/RequestHandler.py", line 270, in __RPCCallFunction\n    uReturnValue = oMethod(*args)\n', '  File "/opt/dirac/pro/DIRAC/WorkloadManagementSystem/Service/JobManagerHandler.py", line 183, in export_submitJob\n    initialMinorStatus=initialMinorStatus)\n', '  File "/opt/dirac/pro/DIRAC/WorkloadManagementSystem/DB/JobDB.py", line 1026, in insertNewJobIntoDB\n    result = jobManifest.load( jdl )\n', '  File "/opt/dirac/pro/DIRAC/WorkloadManagementSystem/Client/JobState/JobManifest.py", line 37, in load\n    return self.loadJDL( dataString )\n', '  File "/opt/dirac/pro/DIRAC/WorkloadManagementSystem/Client/JobState/JobManifest.py", line 45, in loadJDL\n    result = loadJDLAsCFG( jdlString.strip() )\n', '  File "/opt/dirac/pro/DIRAC/Core/Utilities/JDL.py", line 83, in loadJDLAsCFG\n    result = assignValue( key, value, cfg )\n', '  File "/opt/dirac/pro/DIRAC/Core/Utilities/JDL.py", line 54, in assignValue\n    return S_ERROR( "Var %s : %s" % ( key, result[ \'Message\' ] ) )\n'], 'Message': 'Var InputSandbox : value is opened with " but is not closed', 'OK': False, 'rpcStub': (('WorkloadManagement/JobManager', {'delegatedDN': None, 'timeout': 600, 'skipCACheck': False, 'keepAliveLapse': 150, 'delegatedGroup': None}), 'submitJob', ('[ \n    logpar = %(logpar)s;\n    Executable = "$DIRACROOT/scripts/dirac-jobexec";\n    Parameters = 2;\n    exepar = %(exepar)s;\n    JobName = minJob_test__{Ganga_GaudiExec_(0.0)};\n    Priority = 1;\n    Arguments = "jobDescription.xml -o LogLevel=info  -p InputData=%(InputData)s  -p InputSandbox=%(InputSandbox)s";\n    LogLevel = info;\n    InputSandbox = \n        {\n            "%(InputSandbox)s",\n            "SB:ProductionSandboxSE|/SandBox/m/masmith.lhcb_user/09d/6f0/09d6f059566366f147e2cc71053aca51.tar.bz2"\n        };\n    OutputSandbox = \n        {\n            summary.xml,\n            *.root,\n            __postprocesslocations__,\n            Script1_%(logpar)s,\n            std.err,\n            std.out\n        };\n    JobType = User;\n    StdError = std.err;\n    CPUTime = 1209600;\n    JobGroup = lhcb;\n    StdOutput = std.out;\n    Parameters.exepar = \n        {\n            jobScript/masmith_GaudiExec_Job_0.0_26002155-98cf-4e6e-b1a6-d12182407250_script.py,\n            jobScript/masmith_GaudiExec_Job_0.1_26002155-98cf-4e6e-b1a6-d12182407250_script.py\n        };\n    Origin = DIRAC;\n    Parameters.InputData = \n        {\n            {/lhcb/MC/2012/ALLSTREAMS.DST/00042092/0000/00042092_00000057_2.AllStreams.dst},\n            {/lhcb/MC/2012/ALLSTREAMS.DST/00042092/0000/00042092_00000091_2.AllStreams.dst}\n        };\n    Parameters.InputSandbox = \n        {\n            {/afs/cern.ch/user/m/masmith/2910_LFNs.txt,\n            /afs/cern.ch/user/m/masmith/2912_LFNs.txt},\n            {/afs/cern.ch/user/m/masmith/2931_LFNs.txt,\n            /afs/cern.ch/user/m/masmith/2935_LFNs.txt}\n        };\n    Site = ANY;\n    Parameters.logpar = \n        {\n            Ganga_GaudiExec.log,\n            Ganga_GaudiExec.log\n        };\n    MaxCPUTime = 1209600;\n    InputData = %(InputData)s;\n]',)), 'Errno': 0}
```

The submission works if each list has only one file, but multiple files fails. `InputSandbox` seems to be the only one with this problem.

**Comment by fstagni** [comment]
Interesting. So, you are saying that, if you would use
```
j.setParameterSequence("InputSandbox", ['/afs/cern.ch/user/m/masmith/2910_LFNs.txt', '/afs/cern.ch/user/m/masmith/2912_LFNs.txt'], addToWorkflow=True)
```
so a simple list instead of the list of lists that you used, you get no errors?

The job above creates a JDL in which there's the following:
```
...
    InputSandbox = 
        {
            "%(InputSandbox)s",
            "SB:ProductionSandboxSE|/SandBox/m/masmith.lhcb_user/09d/6f0/09d6f059566366f147e2cc71053aca51.tar.bz2"
        };
...
    Parameters.InputSandbox = 
        {
            {/afs/cern.ch/user/m/masmith/2910_LFNs.txt,
            /afs/cern.ch/user/m/masmith/2912_LFNs.txt},
            {/afs/cern.ch/user/m/masmith/2931_LFNs.txt,
            /afs/cern.ch/user/m/masmith/2935_LFNs.txt}
        };
...
```

I am wondering where "SB:ProductionSandboxSE|/SandBox/m/masmith.lhcb_user/09d/6f0/09d6f059566366f147e2cc71053aca51.tar.bz2" comes from. Any idea?

**Comment by andresailer** [comment]
Independent of the error message. I don't think one can use local (or AFS) files in the parametric input sandbox. The parametric input sandbox only works with LFNs.

**Comment by mesmith75** [comment]
So it works with a list of lists if each list only has one entry. I'll try with LFNs...

**Comment by fstagni** [comment]
Hello, news on this?

**Comment by egede** [comment]
@fstagni Mark has spent lots of time trying to get this to work. It only seems to allow a single file in the inputsandbox. This means that we have to pack up things in a tarball for each which will take extra time and cause complications. Together with the max number for each submission, we have to split up things. In our review on Friday of this we concluded that the drawbacks of this in terms of submission were bigger than the benefits, so we do not plan to progress with it at the moment.

The real benefit of a bulk submission would be if, for DaVinci type jobs, the whole dataset could just be given to Dirac and there would then be an asynchronous return of the created subjobs at a later stage. This was in fact the code we had working several years ago but which was not acceptable to LHCb due to some operational reasons.

**Comment by fstagni** [comment]
@rob-c @atsareg Just tagging you here for possible discussions on these points.

**Comment by atsareg** [comment]
Sorry for getting late into this discussion. The bulk job submission mechanism relies on the fact that ALL the jobs in the bunch are sharing the SAME InputSandbox for files uploaded with the jobs (not the LFNs of already uploaded grid files). So, adding a sequence for the InputSandbox does not make sense. What can be the solution ? One way I see is to prepare all the input files for the whole bunch of jobs before the bulk submission (what you do anyway) and put them all in the InputSandbox (not parametrized). The individual jobs then will need a parameterized argument to their executable that says which input sandbox file to use. In this case all the input sandbox files will be loaded for each job and only some of them will be actually used. This is the overhead to be considered. The other limitation is that the total size for this large InputSandbox must not exceed the predefined limit which is 10MB by default.    
On the other hand, I would suggest that Ganga developers check the real need for individual input sandbox files for each job in a bunch. In most cases this is usually a common template with some placeholders replaced by the individual parameters for particular jobs. If this is similar in the Ganga case, you can think of putting individual jobs parameters into the arguments of the job executable rather than into the input sandbox files. 
And one last question: Why splitting the jobs with respect to input data files takes so long time in Ganga ? This looks very strange.

**Comment by mesmith75** [comment]
@atsareg Thanks for the input. This had fallen down my priority list a little bit.

I think in general what you say is right. The actual RTHandlers are currently a bit of a melange, written by different people over different years so I need to spend some time figuring out what they are doing.

As regards the splitter I am not sure what takes the time. It has gotten pretty bad recently with the much larger Run 2 datasets so getting some improvement on that is on my radar.

By the way, is there any progress on the splitter in Dirac? Is that usable?

**Comment by rob-c** [comment]
@atsareg To answer your question about job splitting more fully;

There was the concern that the splitter in DIRAC (I think this still applies but I may be wrong). Matches LFN into subsets based on files with replicas at the same subset of SE.
This became more complicated as portions of the data for different datasets ended up at different subsets of the different combinations of sites possible.
This led to many, many subjobs being created within Ganga which made submitting the jobs an extremely time consuming process.

The solution which was chosen at the time was to be 'smarter' about the splitting.
This involved Ganga requesting all of the replicas for each LFN and then attempting to construct an SE<->CE mapping.
The splitter would then try to create as many larger subsets of LFNs as possible which would reduce the number of (sub)jobs being managed. This was and still is desirable for a few reasons. Ganga on AFS doesn't cope well with numerous large jobs due to various limitations.
The result of this was that jobs would submit with as many of them running over 100LFN as possible rather than many jobs which have a redundancy of being able to run at 3 or 4 or 5 sites due to many replicas being available.

This process could be improved from the Ganga side if Ganga was able to query outright which SE it should use for splitting jobs. @chaen I think you mentioned this was desirable to help manage putting all job output replicas at the same SE. Having knowledge of which SE are expected to work correctly for the next week at the time of submission could improve the situation here as well for users.
Currently Ganga has no automated mechanism to query which SE are already or are going to be in downtime which can lead to jobs being created that can't run despite other replicas for much or all of the data required for that job existing on the grid.

Ganga can now cope better with a large numbers of jobs. (At job submission now this is certainly the case thanks to the ability to bulk-prepare and bulk-submit jobs). Then I suggest that Ganga can attempt to revert back to the old DIRAC splitter which was faster at returning multiple sub-datasets.
(I say suggest as this has not been used in so long I'm certain some dev effort is needed to test/fix the code which does this)
This typically creates many more subjobs per job but this is a tangible problem we're familiar with.
This should mean that this offloads the problem of file replicas being available entirely to DIRAC which feels more correct.
If the DIRAC splitter returns twice the number of subjobs this is likely something Ganga can now cope much better with, but if it returns 10x the number then there will likely be a significant performance trade off elsewhere compared to the current situation.

(Edit, tagging correct Chris)

**Comment by atsareg** [comment]
There are two (at least) different things. One is the splitting algorithm which results in a number of fully specified jobs to submit. The other one is the efficiency of submission of the prepared jobs. The first one requires replica information for all the files in the given dataset (one call to DFC, few secs max) and the SE configuration and status data (one or two calls to the corresponding DIRAC services). The rest is just done in the splitter algorithm locally. So, the whole operation of preparing jobs for submission should take just few seconds for my understanding. Second, job submission to DIRAC using parametric jobs (bulk submission) takes about 1 sec per 100 jobs. I was told that a typical bunch of jobs to submit is about 3000-4000. So, I would expect that the whole operation of splitting the dataset, preparing and submitting jobs should take no more than one minute. Now, if what I have heard is correct, this takes hours. Evidently, there is a room for improvement. So, the whole chain is to be examined for bottle necks.  

About the DIRAC splitters. There are a number of splitters developed for the LHCb production system and, I guess, inherited by Ganga. Those are executed on the client side and all the resulting jobs descriptions are prepared before submission to DIRAC. We considered at some point the possibility to split the jobs on the DIRAC server side (same work done on the server instead of the client). This did not materialize for several reasons. One of them is that in this case the operation of submission becomes asynchronous (you submit a job submission request and then you have to poll the system to follow the request status and finally get back the job IDs and info when the request is processed). This certainly complicates the submission machinery both on the client and on the server side. So, for the moment we stay with just synchronous bulk submission and no splitters on the DIRAC servers side. This proved to work reliably especially after making the whole operation transactional to avoid partial/incomplete job submissions.

**Comment by mesmith75** [comment]
I *think* the current biggest overhead is the splitting. We need to figure out where the time goes with that.

For the submission itself its not so bad currently  - about 10-15 seconds per 100 jobs, of which 5-10 will be overhead from connecting to DIRAC, which will still be there with bulk submission. For sure though we can make some gains there.

Edit: Also the submission is done in parallel so you can multiply 100 by some factor.

**Comment by rob-c** [comment]
Initially time is mainly consumed in many calls to DIRAC to get LFN metadata (replica information) which is only used for splitting.

After that there is always a huge penalty within Ganga while it constructs several thousand class objects which are rarely used. Swapping to use an expensive class object to track every single LFN has proven to not scale well and needs addressing (just based on my experience).

**Comment by atsareg** [comment]
Why "many calls to DIRAC to get LFN metadata" ? DIRAC FC supports bulk queries where you can pass a list of LFNs to get metadata/replicas in one go.

**Comment by rob-c** [comment]
This was broken up because when it used to bulk request the replica info there were timeouts and other issues.
(Edit: In fact the requests are done for 250LFN at a time which works just fine).


That being said the DIRAC overhead is minimal the problems are almost 90% to do with the construction and passing around of several thousand DiracFile objects within Ganga which can be replaced in most places with a flat list of LFN strings. (This was the last time when I threw the problem through valgrind I'm not sure if much has changed)

**Comment by mesmith75** [comment]
@rob-c Yes you are correct. The splitting itself is not optimal but not too bad. I have identified a few areas of improvements but only gaining tens of seconds (the whole thing only takes about a minute for 200 files).

It is the subsequent creation of the ganga job/dataset objects that is taking the time.
